ApeWorX/ape
# feat: support having more than one event with same name

### What I did



fixes: #796 

### How I did it

* Make `_events_` track their events as lists
* Make `__getattr__` raise if `len(_events_[name]) > 1`
* Create method `get_event_by_signature(str) - ContractEvent` that uses the same `_events_` property (only now they are lists) and attempts to find one of those with a matching sig.
* Docs + Tests

### How to verify it

I did this and it worked (but I am a little shocked tbh):

(Interface.sol):

```solidity
interface Interface {
    event FooEvent(uint256 bar);
}
```

(InterfaceImplementation.sol):

```solidity
import "./Interface.sol";


contract InterfaceImplementation is Interface {
    event FooEvent(uint256 bar, uint256 baz);
}
```

Now, you can do this in your projects but if you try the `.` look up of an event, it will error with `AttributeError`. However, if you know the signature as a `str`, you can use the `get_event_by_signature()` method to retrieve the signature.

See the newly added tests as an example.

### Checklist
